sklearn/utils/_metadata_requests.py

```diff
@@ -0,0 +1,1201 @@
+        # the reverse of the MRO since cls is the first in the tuple and object
+        # is the last.
+        defaults = dict()
+        for klass in reversed(inspect.getmro(cls)):
+            klass_defaults = {
+                attr: value
+                for attr, value in vars(klass).items()
+                if "__metadata_request__" in attr
+            }
+            defaults.update(klass_defaults)
+        defaults = dict(sorted(defaults.items()))
+
+        # First take all arguments from the method signatures and have them as
+        # ERROR_IF_PASSED, except X, y, *args, and **kwargs.
+        for method in METHODS:
+            # Here we use `isfunction` instead of `ismethod` because calling `getattr`
+            # on a class instead of an instance returns an unbound function.
+            if not hasattr(cls, method) or not inspect.isfunction(getattr(cls, method)):
+                continue
+            # ignore the first parameter of the method, which is usually "self"
+            params = list(inspect.signature(getattr(cls, method)).parameters.items())[
+                1:
+            ]
+            for pname, param in params:
+                if pname in {"X", "y", "Y"}:
+                    continue
+                if param.kind in {param.VAR_POSITIONAL, param.VAR_KEYWORD}:
+                    continue
+                getattr(requests, method).add_request(
+                    prop=pname,
+                    alias=RequestType.ERROR_IF_PASSED,
+                )
+
+        # Then overwrite those defaults with the ones provided in
+        # __metadata_request__* attributes, which are provided in `requests` here.
+
+        for attr, value in defaults.items():
+            # we don't check for attr.startswith() since python prefixes attrs
+            # starting with __ with the `_ClassName`.
+            substr = "__metadata_request__"
+            method = attr[attr.index(substr) + len(substr) :]
+            for prop, alias in value.items():
+                getattr(requests, method).add_request(prop=prop, alias=alias)
+        return requests
+
+    def _get_metadata_request(self):
+        """Get requested data properties.
+
+        Please check :ref:`User Guide <metadata_routing>` on how the routing
+        mechanism works.
+
+        Returns
+        -------
+        request : MetadataRequest
+            A :class:`~.utils.metadata_requests.MetadataRequest` instance.
+        """
+        if hasattr(self, "_metadata_request"):
+            requests = metadata_router_factory(self._metadata_request)
+        else:
+            requests = self._get_default_requests()
+
+        return requests
+
+    def get_metadata_routing(self):
+        """Get metadata routing of this object.
+
+        Please check :ref:`User Guide <metadata_routing>` on how the routing
+        mechanism works.
+
+        Returns
+        -------
+        routing : dict
+            A dict representing a MetadataRouter.
```

Comment:
It's not obvious to me why we need to return a dict and not a MetadataRouter.

Reply:
Basically third party developers who don't want to have a hard dependency on scikit-learn. Allowing them to return a serialized version makes it easier for them to integrate with us. Although I really doubt if anybody would write their own mechanism, they probably would end up vendoring this file. I'm happy to return the object. It certainly makes things quite a bit faster. ping @agramfort

Reply:
Given the complexity of `sklearn/utils/_metadata_requests.py`, I do not think anyone should (or would) write one themselves and will end up vendoring this file.

The side benefit is that we do not need to invent a serialization format.

Reply:
The PR now returns the object.